### PR TITLE
fluids: Move to local Vec storage and petsc-ops

### LINKED ISSUE #1142: examples/fluids: Use generic MatShell operations for IJacobian
As mentioned in this thread https://github.com/CEED/libCEED/pull/1135#discussion_r1072988454, the new MatShell operations are more-or-less equivalent to what is being done for the IJacobian. So replace `MatMult_NS_IJacobian` with `MatMult_Ceed` and `MatGetDiagonal_NS_IJacobian` with `MatGetDiag_Ceed`. 

Changing them into the new functions reduces redundant code.

### PR DESCRIPTION
**Pull request by jrwrigh**
- Move everything possible over to petsc-ops
- Any continually change data shifted to use Petsc Vec
   -  (Mainly geared at turbulence spanwise statistics)
- Switch IJacobian to use petsc-ops
- Add `CreateMatShell_Ceed` for creating `Mat`
- Move feature-specific setup from `navierstokes.c` to somewhere else (mostly `src/setuplibceed.c`)
- Small cleanups/improvements

Depends on #1149 

Closes #1142

### CODE REVIEW

File: examples/fluids/src/petsc_ops.c
Comment (jrwrigh) [inline examples/fluids/src/petsc_ops.c:438]: Not sure about the usefulness/necessity of this. `MatShellSetVecType` only appears to affect `MatGetVecs`, which we don't use, so possibly a check for nothing.

File: examples/fluids/include/petsc_ops.h
Comment (jrwrigh) [inline examples/fluids/include/petsc_ops.h:25]: Also, thinking about a different name for the struct, as at this point it's starting to look a lot more like a `CeedOperator`, but for Petsc Vecs rather than just a context. Maybe something like `PetscCeedOperator`? Or `FEMOperator` since it can implements the global finite element operator (where `CeedOperator` only implements the local version)?

**Comment by jrwrigh** [comment]
This is ready for review, but obviously depends on #1149, so merging will need to wait until that goes through.

**Comment by jeremylt** [comment]
I like the general look of the changes. I'll wait for #1149 to merge so I can see the changes exclusive to this PR more clearly before digging into any specific comments

**Comment by jrwrigh** [comment]
> I'll wait for #1149 to merge so I can see the changes exclusive to this PR more clearly before digging into any specific comments

The PR is currently set to merge into #1149, so the "File Changed" tab contains the PR-exclusive changes. No worries if you want to wait until #1149 merges though.

**Comment by jeremylt** [comment]
Oh. A big PR into a big PR sort of thing takes a while for me to properly review. I'll find the time, but it may or may not be today.

It can be difficult for sure, but I've been trying to get our PRs down in size where able because its just so difficult to properly review large PRs.

File: examples/fluids/src/petsc_ops.c
Comment (jeremylt) [inline examples/fluids/src/petsc_ops.c:256]: https://petsc.org/main/manualpages/DM/DMCreateLocalVector/

This would set the length and type for you. Unless the DM is just for getting the VecType?

**Review by jeremylt — COMMENTED** [review]
couple of small comments but overall looks good

File: examples/fluids/navierstokes.c
Comment (jrwrigh) [reply examples/fluids/navierstokes.c:21]: That was deliberate to stop the test from running. It's in the[TEMPORARILY IGNORE SGS TEST: SHOULD BE UNDONE BEFORE MERGE](https://github.com/CEED/libCEED/pull/1185/commits/b3c0f218ae3f5a98884cf20a1664faa9526aeb88) commit. That test runs successfully locally, but is waiting approval of posting the neural network parameters before it'll run on CI. I commented it out to make sure everything else still ran at least.

File: examples/fluids/navierstokes.c
Comment (jrwrigh) [reply examples/fluids/navierstokes.c:363]: I had a internal debate about that, but couldn't come to an consensus. It comes down to two main points for me. 

1) Even as we continue to expand this out, it's still an example in the library repo. Exposing the guts in a clear way is a good chunk of it's purpose. I think abstractifying the more periphery features of the code is perfectly reasonable, but it could be argued that the core bits of the code, like this, might benefit from being left alone.

2) More of an issue for IFunction than RHS, but these have some edge cases to them which might warrant leaving them expanded. For example, I'm not sure how to keep the overlapping of `GlobalToLocal` communication for `Q` and `Q_dot` with a `ApplyCeedOperatorGlobalToLocal` call. We could keep it to `ApplyCeedOperatorLocalToLocal`, but then we have abstraction without as much benefit (as far as lines of code removed) while detracting from point 1.

File: examples/fluids/navierstokes.h
Comment (jrwrigh) [reply examples/fluids/navierstokes.h:142]: Same. Stopped myself from doing it in this PR to keep down the scope.

File: examples/fluids/src/petsc_ops.c
Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:67]: I figured `CEED_VECTOR_NONE` is a bit more explicit about intent. Note though that `NULL` actually works just fine. `CeedOperator{Add}Apply` doesn't seem to care between `CEED_VECTOR_NONE` and `NULL`. Documentation-wise, it's supposed to take `CEED_VECTOR_NONE` if there are no active inputs. I think maybe we should have it be exclusively `CEED_VECTOR_NONE` and error if it's passed a `NULL` active vector.

File: examples/fluids/src/petsc_ops.c
Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:73]: Ah, yeah. That is a good idea.

File: examples/fluids/src/petsc_ops.c
Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:256]: DM is just for getting the `VecType`. This is exclusively used for non-DM related `Vec`s. Though I can see how having a DM as an argument could confuse that. I thought about making it an input parameter, but that ends up adding an extra 3 lines to every single call to the function. I guess I could create a 

```c
VecType DMReturnVecType(DM dm) {
  VecType vec_type;
  DMGetVecType(dm, &vec_type);
  return vec_type;
}
```
Then do 
```c
PetscErrorCode CeedOperatorCreateLocalVecs(DM dm, CeedOperator op, Vec *input, Vec *output);

PetscCall(CeedOperatorCreateLocalVecs(DMReturnVecType(dm), op, &input, &output));
```

**Comment by jrwrigh** [comment]
Note I also opened up some threads earlier for discussion in the PR: https://github.com/CEED/libCEED/pull/1185#discussion_r1151264598
https://github.com/CEED/libCEED/pull/1185#discussion_r1151266919

File: examples/fluids/src/petsc_ops.c
Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:67]: > I figured `CEED_VECTOR_NONE` is a bit more explicit about intent.

ie. If I pass `CEED_VECTOR_NONE`, then I'm going to using the `Local` apply function call and pass it a `NULL` local `Vec`:
```c
  PetscCall(OperatorApplyContextCreate(NULL, NULL, ceed, op_quad_coords, CEED_VECTOR_NONE, NULL, NULL, NULL, &op_quad_coords_ctx));
  PetscCall(ApplyCeedOperatorLocalToLocal(NULL, Output_loc, op_quad_coords_ctx));
```

If I pass `NULL`, then I expect that the CeedVector be created on context creation to then be as the container for the local `Vec` I pass to the apply function:
```c
  PetscCall(OperatorApplyContextCreate(NULL, NULL, ceed, op_quad_coords, NULL, NULL, NULL, NULL, &op_quad_coords_ctx));
  PetscCall(ApplyCeedOperatorLocalToLocal(Input_loc, Output_loc, op_quad_coords_ctx));
```

File: examples/fluids/src/petsc_ops.c
Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:260]: I've pushed up changes to the above in fixup commits. If you're good with them, I'll squash them in.

File: examples/fluids/src/petsc_ops.c
Comment (jeremylt) [reply examples/fluids/src/petsc_ops.c:67]: Interesting. We don't have a scenario in Ratel in which there is no input Lvec. It might make sense for clarity to have `ApplyCeedOperatorPassiveToLocal`? 

Could 
```
 if (y_ceed) CeedVectorReferenceCopy(y_ceed, &(*ctx)->y_ceed);
 else if (y_size > 0) CeedVectorCreate(ceed, y_size, &(*ctx)->y_ceed);
```

Its probably a good idea to make `CeedVectorReferenceCopy` work for `CEED_VECTOR_ACTIVE` and `CEED_VECTOR_NONE` (and `CeedVectorDestroy)`.

> I think maybe we should have it be exclusively `CEED_VECTOR_NONE` and error if it's passed a NULL active vector.

Currently we allow `NULL` as an input as a holdover from before `CEED_VECTOR_NONE` existed.

Comment (jeremylt) [reply examples/fluids/src/petsc_ops.c:67]: See https://github.com/CEED/libCEED/pull/1186

File: examples/fluids/src/petsc_ops.c
Comment (jeremylt) [reply examples/fluids/src/petsc_ops.c:99]: https://github.com/CEED/libCEED/pull/1186 would also fix this

File: examples/fluids/src/petsc_ops.c
Comment (jeremylt) [reply examples/fluids/src/petsc_ops.c:260]: Will the DM that was previously passed ever not be of the right size? I suppose it might for some of the operators because the output might not actually have a DM associated with it?

File: examples/fluids/src/petsc_ops.c
Comment (jeremylt) [reply examples/fluids/src/petsc_ops.c:447]: I agree its probably not needed but keeping it is probably not bad either.

File: examples/fluids/problems/newtonian.c
Comment (jeremylt) [reply examples/fluids/problems/newtonian.c:207]: I'm the opposite - negating a multi-part logical statement is more mental overhead for me. I automatically simplify my boolean logic, but at the end of the day its not really any difference one way or another.

File: examples/fluids/navierstokes.c
Comment (jeremylt) [reply examples/fluids/navierstokes.c:363]: I tend to use the wrappers as much as possible since they make sure all the pieces fit together and help me not write some subtle bug.

1 isn't really compelling to me since we have the lower level examples and eventually we don't want anyone to have to write most of this 'glue' code themselves if they are using PETSc.

2 is a more interesting case to me. It makes me think we should merge this work with the operators we can wrap now and revisit in the future the  question of overlapping comm.

File: examples/fluids/include/petsc_ops.h
Comment (jeremylt) [reply examples/fluids/include/petsc_ops.h]: Functional changes now and a better name maybe in the future if there is consensus?

Comment (jrwrigh) [reply examples/fluids/include/petsc_ops.h]: Yeah, sounds fine by me.

File: examples/fluids/navierstokes.c
Comment (jrwrigh) [reply examples/fluids/navierstokes.c:363]: Yeah, sounds good to me.

File: examples/fluids/src/petsc_ops.c
Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:67]: > > I think maybe we should have it be exclusively `CEED_VECTOR_NONE` and error if it's passed a NULL active vector.
> 
> Currently we allow `NULL` as an input as a holdover from before `CEED_VECTOR_NONE` existed.

Gotcha. I'm assuming by holdover, you mean keeping it for backwards compatibility reasons? Otherwise, I think it makes sense to make the user deliberately choose `CEED_VECTOR_NONE` rather than it take `NULL`. 

> It might make sense for clarity to have `ApplyCeedOperatorPassiveToLocal`?

Maybe. I'll have a think on it, as I think `ApplyCeedOperatorLocalToLocal(NULL, Y_loc, ctx);` could work just as well. Currently there's a "backdoor" way to have a `CeedVector` active field; if `NULL` is passed for `X/Y_loc`, it'll just use whatever `ctx->x_ceed` has, which could be an actual `CeedVector` if desired. That backdoor will need to be cutoff for `ApplyCeedOperatorLocalToLocal(NULL,...)` to have the same behavior as a `ApplyCeedOperatorPassiveToLocal`.

File: examples/fluids/src/petsc_ops.c
Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:260]: > Will the DM that was previously passed ever not be of the right size?

Yes. For example, the turbulent statistics are stored locally long-term at quadrature points, not nodes. So there isn't a `DM` that matches with that structure size.

> I suppose it might for some of the operators because the output might not actually have a DM associated with it?

For the turbulent statistics example, this is true for both inputs and outputs, though not at the same time. The collection phase takes in nodal solution field (associated with the 3D solution `DM`) and outputs the statistics at quadrature points (no associated `DM`).

Conversely, the processing step takes those statistics at quadrature points and computes the RHS of an L^2 projection, which is associated with the 2D statistics `DM`.

File: examples/fluids/src/petsc_ops.c
Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:124]: Done.

File: examples/fluids/navierstokes.c
Comment (jrwrigh) [reply examples/fluids/navierstokes.c]: I've now pushed up migrating RHS to the `ApplyCeedOperator*` wrappers. Other than that, the other direct uses of `CeedOperatorApply` are for quadrature point data that is only used as passive inputs for QFunctions.

File: examples/fluids/src/setupts.c
Comment (jrwrigh) [inline examples/fluids/src/setupts.c:254]: Pretty sure this is the correct fix for this, but just pointing out to make sure.

I'm assuming this hasn't bit us previously as we only specify `pmat_pbdiagonal` when `amat_type: shell`.  Otherwise, the A and P matrices are identical and then `CeedOperatorLinearAssemble` is the only one ever called.

File: examples/fluids/src/petsc_ops.c
Comment (jeremylt) [reply examples/fluids/src/petsc_ops.c:118]: > Gotcha. I'm assuming by holdover, you mean keeping it for backwards compatibility reasons?

Yeah, that's why it stuck around for so long. Might be interesting to see in a separate PR what breaks if we remove this.

File: examples/fluids/src/petsc_ops.c
Comment (jeremylt) [inline examples/fluids/src/petsc_ops.c:154]: Note - `PetscFree` only takes `ctx`, so you can use `OperatorApplyContextDestroy(OperatorApplyContext ctx)`  as the function signature. This is the signature expected by `MatShellSetContextDestroy` and similar

Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:154]: > internals of how `MatShellSetContextDestroy` and similar functions are set up doesn't support that.

In what way?

Comment (jeremylt) [reply examples/fluids/src/petsc_ops.c:154]: https://petsc.org/main/manualpages/Mat/MatShellSetContextDestroy/

See the required function signature here

Comment (jrwrigh) [reply examples/fluids/src/petsc_ops.c:154]: Ohhh, I hastily misread/misinterpreted `MatShellSetContextDestroy` as Ratel's equivalent to `OperatorApplyContextDestroy` for some reason. Yeah, that makes sense. I'm guessing is similarly doesn't make sense to have two different destroy functions

Comment (jeremylt) [reply examples/fluids/src/petsc_ops.c:154]: Yeah, I wouldn't make separate functions.

We could go into PETSc and change the signature if we need - I don't think anyone else but us is using those functions. I followed the pattern of `PetscFree`  when creating the functions.

**Review by jeremylt — APPROVED** [review]
This all looks good to me!

**Pull request merged**
